could not load a native library: reactivemongo_netty_transport_native_epoll_<arch>

It seems the `reactivemongo-shaded-native-<os>-<arch>` update from `1.1.0-RC20` to `1.1.0-RC21` (2f2d6489 / https://github.com/lichess-org/lila-ws/pull/899) introduced a name change of the native library (not sure if they intended that or if it is a bug).

In 1.1.0-RC20 it was located at: 
```
x86_64:
/META-INF/native/libreactivemongo_netty_transport_native_epoll_x86_64.so
aarch_64:
/META-INF/native/libreactivemongo_netty_transport_native_epoll_aarch_64.so
```

In 1.1.0-RC21 it is located at:
```
x86_64:
/META-INF/native/libnetty_transport_native_epoll_x86_64.so
aarch_64:
/META-INF/native/libnetty_transport_native_epoll_aarch_64.so
```

It seems lila-ws still expects the old name while trying to find the library: 
```
Caused by: java.lang.UnsatisfiedLinkError: could not load a native library: reactivemongo_netty_transport_native_epoll_aarch_64
        at reactivemongo.io.netty.util.internal.NativeLibraryLoader.load(NativeLibraryLoader.java:244)
```

<details>
<summary>full stacktrace</summary>

```
[ERROR] [08/16/2026 16:23:29.201] [reactivemongo-akka.actor.internal-dispatcher-3] [akka://reactivemongo/user/Connection-2] failed to load the required native
library
akka.actor.ActorInitializationException: akka://reactivemongo/user/Connection-2: exception during creation
        at akka.actor.ActorInitializationException$.apply(Actor.scala:196)
        at akka.actor.ActorCell.create(ActorCell.scala:664)
        at akka.actor.ActorCell.invokeAll$1(ActorCell.scala:514)
        at akka.actor.ActorCell.systemInvoke(ActorCell.scala:536)
        at akka.dispatch.Mailbox.processAllSystemMessages(Mailbox.scala:295)
        at akka.dispatch.Mailbox.run(Mailbox.scala:230)
        at akka.dispatch.Mailbox.exec(Mailbox.scala:243)
        at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:511)
        at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1450)
        at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:2019)
        at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:187)
Caused by: java.lang.reflect.InvocationTargetException
        at java.base/jdk.internal.reflect.DirectConstructorHandleAccessor.newInstance(DirectConstructorHandleAccessor.java:74)
        at java.base/java.lang.reflect.Constructor.newInstanceWithCaller(Constructor.java:499)
        at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:483)
        at reactivemongo.core.netty.Pack$.$anonfun$4(Pack.scala:97)
        at reactivemongo.core.netty.ChannelFactory.<init>(ChannelFactory.scala:51)
        at reactivemongo.core.actors.StandardDBSystem.newChannelFactory(MongoDBSystem.scala:2160)
        at reactivemongo.core.actors.MongoDBSystem.preStart(MongoDBSystem.scala:398)
        at reactivemongo.core.actors.MongoDBSystem.preStart$(MongoDBSystem.scala:94)
        at reactivemongo.core.actors.StandardDBSystem.preStart(MongoDBSystem.scala:2150)
        at akka.actor.Actor.aroundPreStart(Actor.scala:548)
        at akka.actor.Actor.aroundPreStart$(Actor.scala:471)
        at reactivemongo.core.actors.StandardDBSystem.aroundPreStart(MongoDBSystem.scala:2150)
        at akka.actor.ActorCell.create(ActorCell.scala:644)
        ... 9 more
Caused by: java.lang.UnsatisfiedLinkError: failed to load the required native library
        at reactivemongo.io.netty.channel.epoll.Epoll.ensureAvailability(Epoll.java:90)
        at reactivemongo.io.netty.channel.epoll.EpollIoHandler.newFactory(EpollIoHandler.java:107)
        at reactivemongo.io.netty.channel.epoll.EpollEventLoopGroup.<init>(EpollEventLoopGroup.java:123)
        at reactivemongo.io.netty.channel.epoll.EpollEventLoopGroup.<init>(EpollEventLoopGroup.java:110)
        at reactivemongo.io.netty.channel.epoll.EpollEventLoopGroup.<init>(EpollEventLoopGroup.java:87)
        at reactivemongo.io.netty.channel.epoll.EpollEventLoopGroup.<init>(EpollEventLoopGroup.java:63)
        at reactivemongo.io.netty.channel.epoll.EpollEventLoopGroup.<init>(EpollEventLoopGroup.java:56)
        at java.base/jdk.internal.reflect.DirectConstructorHandleAccessor.newInstance(DirectConstructorHandleAccessor.java:62)
        ... 21 more
Caused by: java.lang.UnsatisfiedLinkError: could not load a native library: reactivemongo_netty_transport_native_epoll_aarch_64
        at reactivemongo.io.netty.util.internal.NativeLibraryLoader.load(NativeLibraryLoader.java:244)
        at reactivemongo.io.netty.channel.epoll.Native.loadNativeLibrary(Native.java:334)
        at reactivemongo.io.netty.channel.epoll.Native.<clinit>(Native.java:96)
        at reactivemongo.io.netty.channel.epoll.Epoll.<clinit>(Epoll.java:42)
        ... 28 more
```
</details>

https://repo1.maven.org/maven2/org/reactivemongo/reactivemongo-shaded-native-linux-aarch-64/1.1.0-RC20/
https://repo1.maven.org/maven2/org/reactivemongo/reactivemongo-shaded-native-linux-aarch-64/1.1.0-RC21/

https://repo1.maven.org/maven2/org/reactivemongo/reactivemongo-shaded-native-linux-x86-64/1.1.0-RC20/
https://repo1.maven.org/maven2/org/reactivemongo/reactivemongo-shaded-native-linux-x86-64/1.1.0-RC21/

--- a/build.sbt
+++ b/build.sbt
@@ -10,9 +10,13 @@ inThisBuild(
 )
 
 Universal / target := baseDirectory.value / "target" / "universal"
-val os = if (sys.props.get("os.name").exists(_.startsWith("Mac"))) "osx" else "linux"
-val arch = if (sys.props.get("os.arch").exists(_.startsWith("aarch64"))) "aarch-64" else "x86-64"
-val arch_ = arch.replace("-", "_")
+
+def nettyTransport =
+  val arch = if (sys.props.get("os.arch").exists(_.startsWith("aarch64"))) "aarch_64" else "x86_64"
+  val (os, notifier) =
+  if System.getProperty("os.name").toLowerCase.startsWith("mac") then ("osx", "kqueue")
+  else ("linux", "epoll")
+  ("io.netty" % s"netty-transport-native-$notifier" % nettyVersion).classifier(s"$os-$arch")
 
 val pekkoVersion = "1.6.0"
 val kamonVersion = "2.8.1"
@@ -33,10 +37,7 @@ lazy val `lila-ws` = project
       "io.lettuce" % "lettuce-core" % "7.6.0.RELEASE",
       "io.netty" % "netty-handler" % nettyVersion,
       "io.netty" % "netty-codec-http" % nettyVersion,
-      ("io.netty" % s"netty-transport-native-epoll" % nettyVersion)
-        .classifier(s"linux-$arch_"),
-      ("io.netty" % s"netty-transport-native-kqueue" % nettyVersion)
-        .classifier(s"osx-$arch_"),
+      nettyTransport,
       "com.github.lichess-org.scalalib" %% "scalalib-lila" % "11.10.11",
       "com.github.lichess-org.scalachess" %% "scalachess" % chessVersion,
       "com.github.lichess-org.scalachess" %% "scalachess-play-json" % chessVersion,

--- a/build.sbt
+++ b/build.sbt
@@ -28,9 +28,7 @@ lazy val `lila-ws` = project
     resolvers += "jitpack".at("https://jitpack.io"),
     resolvers += "lila-maven".at("https://raw.githubusercontent.com/ornicar/lila-maven/master"),
     libraryDependencies ++= Seq(
-      ("org.reactivemongo" %% "reactivemongo" % "1.1.0-RC20")
-        .exclude("org.scala-lang.modules", "scala-java8-compat_2.13"),
-      "org.reactivemongo" % s"reactivemongo-shaded-native-$os-$arch" % "1.1.0-RC20",
+      "org.reactivemongo" %% "reactivemongo" % "1.1.0-pekko.noshaded.RC21",
       "io.lettuce" % "lettuce-core" % "7.6.0.RELEASE",
       "io.netty" % "netty-handler" % nettyVersion,
       "io.netty" % "netty-codec-http" % nettyVersion,

--- a/src/main/scala/netty/NettyServer.scala
+++ b/src/main/scala/netty/NettyServer.scala
@@ -4,8 +4,6 @@ package netty
 import com.typesafe.config.Config
 import com.typesafe.scalalogging.Logger
 import io.netty.bootstrap.ServerBootstrap
-import io.netty.channel.epoll.{ EpollEventLoopGroup, EpollServerSocketChannel }
-import io.netty.channel.kqueue.{ KQueueEventLoopGroup, KQueueServerSocketChannel }
 import io.netty.channel.{ Channel, ChannelInitializer }
 import io.netty.handler.codec.http.*
 
@@ -17,10 +15,9 @@ final class NettyServer(
 )(using Executor, Scheduler):
   private val logger = Logger(getClass)
   private val threads = config.getInt("netty.threads")
-  private val (parent, workers, channelClass) =
-    if System.getProperty("os.name").toLowerCase.startsWith("mac") then
-      (new KQueueEventLoopGroup(1), new KQueueEventLoopGroup(threads), classOf[KQueueServerSocketChannel])
-    else (new EpollEventLoopGroup(1), new EpollEventLoopGroup(threads), classOf[EpollServerSocketChannel])
+  private val pack = Pack.instance
+  private val parent = pack.eventLoopGroupFactory(1)
+  private val workers = pack.eventLoopGroupFactory(threads)
   private val connector = ActorChannelConnector(clients, config, settings, workers)
 
   def start(): Unit =
@@ -32,7 +29,7 @@ final class NettyServer(
       val boot = new ServerBootstrap
       boot
         .group(parent, workers)
-        .channel(channelClass)
+        .channel(pack.channelClass)
         .childHandler(
           new ChannelInitializer[Channel]:
             override def initChannel(ch: Channel): Unit =

--- a/src/main/scala/netty/Pack.scala
+++ b/src/main/scala/netty/Pack.scala
@@ -1,0 +1,54 @@
+package lila.ws
+package netty
+
+import io.netty.channel.ServerChannel
+import io.netty.channel.EventLoopGroup
+import scala.util.control.NonFatal
+
+// Adapted from https://github.com/ReactiveMongo/ReactiveMongo/blob/5560470ee409da827feee161c16631042e194263/driver/src/main/scala/core/netty/Pack.scala#L15
+private final class Pack(
+    val eventLoopGroupFactory: Int => EventLoopGroup,
+    val channelClass: Class[? <: ServerChannel]
+)
+
+private object Pack:
+  private val kqueuePkg: String = "io.netty.channel.kqueue"
+  private val epollPkg: String = "io.netty.channel.epoll"
+
+  def instance: Pack =
+    epoll.orElse(kqueue).getOrElse { throw RuntimeException("Can't initialize either Netty Epoll or Kqueue") }
+
+  private def epoll: Option[Pack] =
+    try
+      Some(Class.forName(s"${epollPkg}.EpollServerSocketChannel")).map { cls =>
+        val chanClass = cls.asInstanceOf[Class[? <: ServerChannel]]
+        val groupClass = Class
+          .forName(s"${epollPkg}.EpollEventLoopGroup")
+          .asInstanceOf[Class[? <: EventLoopGroup]]
+
+        val groupCtor = groupClass.getDeclaredConstructor(classOf[Int])
+        new Pack(
+          nThreads => groupCtor.newInstance(Int.box(nThreads)),
+          chanClass
+        )
+      }
+    catch
+      case NonFatal(cause) =>
+        None
+
+  private def kqueue: Option[Pack] =
+    try
+      Some(Class.forName(s"${kqueuePkg}.KQueueServerSocketChannel")).map { cls =>
+        val chanClass = cls.asInstanceOf[Class[? <: ServerChannel]]
+        val groupClass = Class
+          .forName(s"${kqueuePkg}.KQueueEventLoopGroup")
+          .asInstanceOf[Class[? <: EventLoopGroup]]
+        val groupCtor = groupClass.getDeclaredConstructor(classOf[Int])
+        new Pack(
+          nThreads => groupCtor.newInstance(Int.box(nThreads)),
+          chanClass
+        )
+      }
+    catch
+      case NonFatal(cause) =>
+        None

--- a/src/main/scala/netty/Pack.scala
+++ b/src/main/scala/netty/Pack.scala
@@ -1,8 +1,8 @@
 package lila.ws
 package netty
 
-import io.netty.channel.ServerChannel
-import io.netty.channel.EventLoopGroup
+import io.netty.channel.{ EventLoopGroup, ServerChannel }
+
 import scala.util.control.NonFatal
 
 // Adapted from https://github.com/ReactiveMongo/ReactiveMongo/blob/5560470ee409da827feee161c16631042e194263/driver/src/main/scala/core/netty/Pack.scala#L15
@@ -33,7 +33,7 @@ private object Pack:
         )
       }
     catch
-      case NonFatal(cause) =>
+      case NonFatal(_) =>
         None
 
   private def kqueue: Option[Pack] =
@@ -50,5 +50,5 @@ private object Pack:
         )
       }
     catch
-      case NonFatal(cause) =>
+      case NonFatal(_) =>
         None

